README.md
# pacs <a href='https://github.com/polkas/pacs'><img src='man/figures/pacs_logo.png' align="right" width="200px" /></a>
[![R build status](https://github.com/polkas/pacs/workflows/R-CMD-check/badge.svg)](https://github.com/polkas/pacs/actions)
[![CRAN](https://www.r-pkg.org/badges/version/pacs)](https://cran.r-project.org/package=pacs)
[![codecov](https://codecov.io/gh/Polkas/pacs/branch/master/graph/badge.svg)](https://app.codecov.io/gh/Polkas/pacs)

[**For more information please visit the pacs website**](https://polkas.github.io/pacs/).

[**Functions Reference**](https://polkas.github.io/pacs/reference/index.html)

A set of tools that make life easier for developers and maintainers of R packages.

- Validating the library, packages and `renv` lock files.
- Exploring complexity of a certain package like evaluating its size in bytes with dependencies.
- The shiny app complexity could be explored too.
- Assessing the life duration of a specific package version.
- Checking a CRAN package check page status for any errors and warnings.
- Retrieving a DESCRIPTION or NAMESPACE file for any package version.
- Comparing DESCRIPTION or NAMESPACE files between different package versions.
- Getting a list of all releases for a specific package.
- The Bioconductor is partly supported.

**An Internet connection is required to take full advantage of most of the features.**

**Almost all calls which requiring an Internet connection are cached (for 30 minutes) by the `memoise` package, so the second invocation of the same command (and arguments) is immediate. Restart the R session if you want to clear cached data.**

## Installation

Development version:

```r
# install.packages("remotes")
remotes::install_github("polkas/pacs")
```

CRAN:

```r
install.packages("pacs")
```

## Examples

For more examples with descriptions please visit [**the walk-through vignette**](https://polkas.github.io/pacs/articles/GettingStarted.html).

Validate the library:

```r
pacs::lib_validate(
  checkred = list(
    scope = c("ERROR", "FAIL"),
    # by default all flavors
    flavors = NULL
  )
)
```

All package check statuses for all CRAN servers:

```r
pacs::checked_packages()
```

All CRAN releases of a package:

```r
pacs::pac_timemachine("dplyr")
```

Compare package DESCRIPTION file dependencies between versions:

```r
pacs::pac_compare_versions("shiny", "1.0.0", "1.5.0")
```

Looking for heavy dependencies:

```r
pacs::pac_deps_heavy("caret")
```


NEWS.md
# pacs 0.5.1

* due to instability issues with "crandb", the lifeduration and timemachine functions have been adjusted to fetch data directly from CRAN by default.
* protect against 400 results from crandb. Please use new options if needed `options(pacs.crandb_ntry = 3)` and `options(pacs.crandb_nsleep = 0.1)` to control crandb fetch.
. Linked with https://github.com/r-hub/pkgsearch/issues/117#issuecomment-1651081139
* add informative messages when a web call failed.
* small improvements in the tinyverse vignette.

# pacs 0.4.10

* remove the not needed testthat log file.

# pacs 0.4.9

* update all `*_validate` functions to treat the analyzed version as newest one if is bigger or equal to the newest CRAN, not only equal as before.
* improved quality of vignettes.
* add usage of the `pacs::checked_packages` function in README.

# pacs 0.4.8

* fix the "how-r-searches-and-finds-stuff" url.
* a new `match_flavors` function to get CRAN server names matched to local `OS`.
* improving docs.

# pacs 0.4.6

* a new function `pac_deps_heavy`, the demand revealed during the `useR` 2022 Conference. The direct package dependencies and counts of dependencies for each of them. Help to identify heavy packages.
* small update of `tinyverse` vignette.
* fix the documentation, `repos` argument in 2 functions get part of `base` argument description.
* rm some of `Suggests` packages.

# pacs 0.4.5

* `pac_deps_dev` and `pac_deps_user` could be used for local packages now.
* a new optional `lib_validate` column inspired by Winston Chang, if the packages have a proper `Built` version.
* a poster vignette.

# pacs 0.4.4

* `tinyverse` vignette.
* fixed edge cases for `pacs::pac_deps`.
* fixed `lock_validate`, the condition for validating the maximum number of packages was taken from the library size not the lock file size.
* two new higher-level functions `pacs::pac_deps_user` and `pacs::pac_deps_dev`.
* `renv` transferred to `Suggests`, as used by only 2 functions.
* improved documentation.
* updated pkgdown website.

# pacs 0.4.3

* added an additional source for getting all releases of packages, the `metacran` db (new default).
* added a new validation function, `lock_validate`. It helps to validate a `renv` lock file.
* more efficient check of the Internet connection with the `curl` package.
* more elegant separation of the code.
* more `testthat` tests, `mockery`.

# pacs 0.4.2

* fix one of the validation of the `lib_validate` function.
* cache results with `memoise` for 30 minutes, not 1 hour.
* small README update.

# pacs 0.4.1

* `pacs::biocran_repos` not includes the `BioCbooks` field for `Bioc` versions smaller than "3.12".
* improve `pac_deps`, NA versions for not installed packages.
* updated `fields` argument of the`pac_deps` function, now has the same shape as `which` argument in the `tools::package_dependencies` function.
* more tests, 90 percent coverage.
* More clear distinction when working without Internet.

# pacs 0.4.0

* `pac_deps` fully depends on `tools::package_dependencies`.
* fix `pac_deps` problems when Suggests field is used, showed only installed packages.
* expose the `dir_size` function.
* add new functions `app_size` and `app_deps`.
* "getting started" vignette instead of the long README.
* get back functions reference table in the README.
* new Imports dependency, `renv`.

# pacs 0.3.8

* fix testthat tests.
* skip an unstable test on CRAN, a new package could be installed on the server during the check is running.
* accepting `"Enhances"` field in all related functions.
* improve the code for reading html tables.
* updated README file.

# pacs 0.3.7

* fixed a bug for `lib_validate` when a custom library path is used.
* updated the default value of the `lib.loc` argument from `NULL` to `.libPaths()`, as it will be more suited when using `withr` package. 
* updated README file.
* accepting `"Enhances"` field when checking dependencies.

# pacs 0.3.6 

* fix testthat tests to be more stable when dependency packages are updated.

# pacs 0.3.5

* Fixed `pac_timemachine` when from and to arguments are used and to is bigger than last Archived Date.
* `pac_timemachine` has a well ordered `rownames` now.
* Fixed `pac_compare_namespace`, how S3methods are handled.
* Fixed `pac_health` and related when the at argument is used and a package is installed locally.

# pacs 0.3.4

* `pac_deps` might show duplicated values if package was installed under a few different `libPaths`, now it is fixed.
* `lib_validate` will show packages which are required by DESCRIPTION file and not exist in `installed.packages`. This might occur when e.g. `install.packages` is corrupted. `pac_validate` already has such functionality.
* Update `roxygen2` of `pac_comapre_namespace` as the note field wrongly suggested that it works only for exports.
* remove "packages versions" section from README file.
* `pac_validate` will return `data.frame` with additional column "direct", which will point the direct dependency from DESCRIPTION file.
* Improve testthat tests for `pac_validate`.
* Improve README file for `lib_validate`.

# pacs 0.3.3

* Partly support BioConductor.
* Default repositories mainly now consists of not only CRAN as BioConductor is considered too.
* Fix `lib_validate` as it not take into account statuses for not fully performed tests, it was case for less than 1% of all packages.
* Add note that `checked_packages` could return duplicated observations which is expected.
* Improve merge in `lib_validate` to show non-CRAN packages too.
* Rewrite some tests as some R CRAN servers might shut down.
* Polishing code of `pac_lifeduration`.

# pacs 0.3.2

* New CRAN release.
* Added `\dontrun` to most of examples as `utils::installed.packages` consumes 30 seconds on win-builder.

# pacs 0.3.1

* Small docs improvements.
* Additional tests, maintain coverage over 80%.
* Additional input validation.
* Added `xml2` to imports.
* Update `README` file.

# pacs 0.3.0

* Added `pac_checkpage` and `cran_flavors`.

R/utils.R
replaceNA <- function(vec, with) {
  vec[is.na(vec)] <- with
  vec
}

isNA <- function(x) {
  isTRUE(length(x) == 1 && is.na(x))
}

#' Maximum version across the vector
#' @description Reduce function over the `utils::compareVersion`
#' @inheritParams standard_args
#' @return `character` maximum version
#' @examples
#' compareVersionsMax(c("1.1.1", "0.2.0"))
#' @export
#'
compareVersionsMax <- function(vec, na.rm = TRUE) {
  stopifnot(is.logical(na.rm))
  if (length(vec) == 1) {
    return(vec)
  }
  if (na.rm) vec <- stats::na.omit(vec)
  Reduce(
    function(x, y) {
      cc <- utils::compareVersion(x, y)
      if (cc == 1) {
        x
      } else if (cc == -1) {
        y
      } else {
        x
      }
    },
    vec
  )
}

#' Minimum version across the vector
#' @description Reduce function over the `utils::compareVersion`
#' @inheritParams standard_args
#' @return `character` minimal version
#' @examples
#' compareVersionsMin(c("1.1.1", "0.2.0"))
#' @export
#'
compareVersionsMin <- function(vec, na.rm = TRUE) {
  stopifnot(is.logical(na.rm))
  if (length(vec) == 1) {
    return(vec)
  }
  if (na.rm) vec <- stats::na.omit(vec)
  Reduce(
    function(x, y) {
      cc <- utils::compareVersion(x, y)
      if (cc == 1) {
        y
      } else if (cc == -1) {
        x
      } else {
        x
      }
    },
    vec
  )
}

#' Size of the package
#' @description size of package.
#' @inheritParams standard_args
#' @return `numeric` size in bytes, to get MB ten divide by `10**6`.
#' @export
#' @examples
#' \dontrun{
#' cat(pacs::dir_size(system.file(package = "stats")) / 10**6, "MB")
#' }
dir_size <- function(path = ".", recursive = TRUE) {
  stopifnot(is.character(path))
  stopifnot(is.logical(recursive))
  files <- list.files(path, full.names = T, recursive = recursive)
  vect_size <- sapply(files, function(x) file.size(x))
  size_files <- sum(vect_size)
  size_files
}

is_online <- function() {
  curl::has_internet()
}

#' Get base R packages
#' @description get base packages, all or only `startup`.
#' @inheritParams standard_args
#' @return `character` vector
#' @examples
#' \dontrun{
#' pacs_base()
#' pacs_base(startup = TRUE)
#' }
#' @export
#'
pacs_base <- function(startup = FALSE) {
  stopifnot(is.logical(startup))
  if (startup) {
    c(getOption("defaultPackages"), "base")
  } else {
    pacs_base_all()
  }
}

pacs_base_all_raw <- function() {
  rownames(installed_packages(lib.loc = NULL, priority = "base"))
}

pacs_base_all <- memoise::memoise(pacs_base_all_raw)

installed_descriptions <- function(lib.loc, fields, deps = NULL) {
  installed_agg <- installed_agg_fun(lib.loc, fields)

  paks <- installed_agg[, fields, drop = FALSE]

  if (!is.null(deps)) {
    nams <- installed_agg[, c("Package")]
    paks <- paks[nams %in% deps, , drop = FALSE]
  }

  joint_cols <- apply(paks, 1, function(x) paste(x, collapse = ","))

  desc_e <- extract_deps(joint_cols)

  packages <- desc_e$packages
  versions <- desc_e$versions

  joint <- data.frame(
    Version = unlist(lapply(seq_along(packages), function(x) replaceNA(versions[[x]], ""))),
    Package = unlist(lapply(seq_along(packages), function(x) replace(packages[[x]], packages[[x]] == "NA", NA))),
    stringsAsFactors = FALSE
  )

  res_agg <- stats::aggregate(
    joint[, c("Version"), drop = FALSE],
    list(Package = joint$Package),
    compareVersionsMax
  )

  res_agg$Version[is.na(res_agg$Version)] <- ""

  res_agg[!is.na(res_agg$Package), ]
}

installed_agg_fun_raw <- function(lib.loc, fields) {
  installed_df <- as.data.frame(installed_packages(lib.loc = lib.loc))
  installed_agg <- stats::aggregate(
    installed_df[, c("Version", fields), drop = FALSE],
    list(Package = installed_df$Package),
    function(x) x[1]
  )
  installed_agg
}

installed_agg_fun <- memoise::memoise(installed_agg_fun_raw, cache = cachem::cache_mem(max_age = 30 * 60))

#' List Available Packages at CRAN-like Repositories
#' @description available_packages returns a matrix of details corresponding to packages currently available at one or more repositories. The current list of packages is downloaded over the internet (or copied from a local mirror).
#' @param repos character vector URLs of the repositories to use. By default checking CRAN and newest Bioconductor per R version. Default `pacs::biocran_repos()`
#' @keywords internal
available_packages <- function(repos) {
  tryCatch(available_packages_raw(repos = repos), error = function(e) NA)
}

available_packages_raw <- memoise::memoise(utils::available.packages, cache = cachem::cache_mem(max_age = 30 * 60))

installed_packages <- function(lib.loc, priority = NULL) {
  installed_packages_raw(lib.loc = lib.loc, priority = priority)
}

installed_packages_raw <- memoise::memoise(utils::installed.packages, cache = cachem::cache_mem(max_age = 30 * 60))

extract_deps <- function(x) {
  splited <- stri_split_fixed(x, ",", )
  trimed <- lapply(splited, stri_trim)
  v_reg <- function(x) {
    vapply(
      stringi::stri_match_all(x, regex = "([0-9\\.-]+)\\)"),
      function(i) `[`(i, 2),
      character(1)
    )
  }
  versions <- lapply(trimed, v_reg)
  v_pac <- function(x) {
    vapply(
      stri_split_regex(x, "[ \n\\(]"),
      function(x) `[[`(x, 1),
      character(1)
    )
  }
  pacs <- lapply(trimed, v_pac)
  list(packages = pacs, versions = versions)
}

available_descriptions <- function(repos, fields, deps = NULL) {
  available_agg <- available_agg_fun(repos, fields)

  paks <- available_agg[, fields, drop = FALSE]

  if (!is.null(deps)) {
    nams <- available_agg[, c("Package")]
    paks <- paks[nams %in% deps, , drop = FALSE]
  }

  joint_cols <- apply(paks, 1, function(x) paste(x, collapse = ","))

  desc_e <- extract_deps(joint_cols)

  packages <- desc_e$packages
  versions <- desc_e$versions

  joint <- data.frame(
    Version = unlist(lapply(seq_along(packages), function(x) replaceNA(versions[[x]], ""))),
    Package = unlist(lapply(seq_along(packages), function(x) replace(packages[[x]], packages[[x]] == "NA", NA))),
    stringsAsFactors = FALSE
  )

  res_agg <- stats::aggregate(
    joint[, c("Version"), drop = FALSE],
    list(Package = joint$Package),
    compareVersionsMax
  )

  res_agg$Version[is.na(res_agg$Version)] <- ""

  res_agg
}

available_agg_fun_raw <- function(repos, fields) {
  available_df <- as.data.frame(available_packages(repos = repos))
  available_agg <- stats::aggregate(
    available_df[, c("Version", fields), drop = FALSE],
    list(Package = available_df$Package),
    function(x) x[1]
  )
  available_agg
}

available_agg_fun <- memoise::memoise(available_agg_fun_raw, cache = cachem::cache_mem(max_age = 30 * 60))

expand_dependency <- function(fields) {
  if (length(fields) == 1 && fields %in% c("strong", "all", "most")) {
    switch(fields,
      strong = c("Depends", "Imports", "LinkingTo"),
      most = c("Depends", "Imports", "LinkingTo", "Suggests"),
      all = c("Depends", "Imports", "LinkingTo", "Suggests", "Enhances")
    )
  } else {
    stopifnot(all(fields %in% c("Depends", "Imports", "Suggests", "LinkingTo", "Enhances")))
    fields
  }
}

cran_archive_file <- function(pac, version, repos, file) {
  last_version <- pac_last(pac, repos)

  if (isTRUE(!is.null(version) && version != last_version)) {
    base_url <- sprintf("https://cran.r-project.org/src/contrib/Archive/%s", pac)
  } else {
    base_url <- "https://cran.r-project.org/src/contrib"
    version <- last_version
  }

  d_url <- sprintf(
    "%s/%s_%s.tar.gz",
    base_url,
    pac,
    version
  )

  temp_tar <- tempfile(fileext = ".tar.gz")

  download <- try(
    {
      suppressWarnings(utils::download.file(d_url,
        destfile = temp_tar,
        quiet = TRUE
      ))
    },
    silent = TRUE
  )

  if (inherits(download, "try-error")) {
    result <- structure(list(), package = pac, version = version)
  } else {
    temp_dir <- tempdir()
    utils::untar(temp_tar, exdir = temp_dir)
    # tabs are not acceptable
    result <- switch(file,
      DESCRIPTION = as.list(read.dcf(file.path(temp_dir, pac, "DESCRIPTION"))[1, ]),
      NAMESPACE = readLines(file.path(temp_dir, pac, "NAMESPACE"), warn = FALSE)
    )
  }
  unlink(temp_tar)
  result
}

read_html_table <- function(table_lines) {
  rr_range <- grep("</?table[^>]*>", table_lines)
  if (length(rr_range) != 2) {
    return(NA)
  }
  rrr <- table_lines[(rr_range[1] + 1):(rr_range[2] - 1)]
  rrr_all <- paste(rrr, collapse = "\n")
  rrr_html <- read_html(rrr_all)
  list(html = rrr_html, lines = rrr)
}

crandb_json <- function(packages,
                        limit = getOption("pacs.crandb_limit", 100),
                        ntry = getOption("pacs.crandb_ntry", 3),
                        nsleep = getOption("pacs.crandb_nsleep", 0.1)) {
  if (!is_online()) {
    message("No internet connection detected.\n")
    return(NA)
  }

  crandb_url <- sprintf(
    'https://crandb.r-pkg.org/-/allall?keys=["%s"]&limit=%s',
    paste(packages, collapse = '","'),
    limit
  )

  result <- NA
  for (iter in seq_len(ntry)) {
    fetch_call <- try(curl::curl_fetch_memory(crandb_url), silent = TRUE)

    if (!inherits(fetch_call, "try-error") && (fetch_call$status_code == 200)) {
      result <- jsonlite::fromJSON(rawToChar(fetch_call$content))
      break
    }

    Sys.sleep(nsleep)
  }

  result
}
